File: samples/cext-trey-research-auth/src/functions/projects.ts
/* This code sample provides a starter kit to implement server side logic for your Teams App in TypeScript,
 * refer to https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference for complete Azure Functions
 * developer guide.
 */

import { app, HttpRequest, HttpResponseInit, InvocationContext } from "@azure/functions";
import ProjectApiService from "../services/ProjectApiService";
import { ApiProject, ApiAddConsultantToProjectResponse, ErrorResult } from "../model/apiModel";
import { HttpError, cleanUpParameter } from "../services/Utilities";
import IdentityService from "../services/IdentityService";

/**
 * This function handles the HTTP request and returns the project information.
 *
 * @param {HttpRequest} req - The HTTP request.
 * @param {InvocationContext} context - The Azure Functions context object.
 * @returns {Promise<Response>} - A promise that resolves with the HTTP response containing the project information.
 */

// Define a Response interface.
interface Response extends HttpResponseInit {
    status: number;
    jsonBody: {
        results: ApiProject[] | ApiAddConsultantToProjectResponse | ErrorResult;
    };
}
export async function projects(
    req: HttpRequest,
    context: InvocationContext
): Promise<Response> {
    context.log("HTTP trigger function projects processed a request.");
    // Initialize response.
    const res: Response = {
        status: 200,
        jsonBody: {
            results: [],
        },
    };

    try {

        // Will throw an exception if the request is not valid
        const userInfo = await IdentityService.validateRequest(req);

        const id = req.params?.id?.toLowerCase();
        let body = null;
        switch (req.method) {
            case "GET": {

                let projectName = req.query.get("projectName")?.toString().toLowerCase() || "";
                let consultantName = req.query.get("consultantName")?.toString().toLowerCase() || "";

                console.log(`➡️ GET /api/projects: request for projectName=${projectName}, consultantName=${consultantName}, id=${id}`);

                projectName = cleanUpParameter("projectName", projectName);
                consultantName = cleanUpParameter("consultantName", consultantName);

                if (id) {
                    const result = await ProjectApiService.getApiProjectById(id);
                    res.jsonBody.results = [result];
                    console.log(`   ✅ GET /api/projects: response status ${res.status}; 1 projects returned`);
                    return res;
                }

                // Use current user if the project name is user_profile
                if (projectName.includes('user_profile')) {
                    const result = await ProjectApiService.getApiProjects("", userInfo.name);
                    res.jsonBody.results = result;
                    console.log(`   ✅ GET /api/projects for current user response status ${res.status}; ${result.length} projects returned`);
                    return res;
                }

                const result = await ProjectApiService.getApiProjects(projectName, consultantName);
                res.jsonBody.results = result;
                console.log(`   ✅ GET /api/projects: response status ${res.status}; ${result.length} projects returned`);
                return res;
            }
            case "POST": {
                switch (id.toLocaleLowerCase()) {
                    case "assignconsultant": {
                        try {
                            const bd = await req.text();
                            body = JSON.parse(bd);
                        } catch (error) {
                            throw new HttpError(400, `No body to process this request.`);
                        }
                        if (body) {
                            const projectName = cleanUpParameter("projectName", body["projectName"]);
                            if (!projectName) {
                                throw new HttpError(400, `Missing project name`);
                            }
                            const consultantName = cleanUpParameter("consultantName", body["consultantName"]?.toString() || "");
                            if (!consultantName) {
                                throw new HttpError(400, `Missing consultant name`);
                            }
                            const role = cleanUpParameter("Role", body["role"]);
                            if (!role) {
                                throw new HttpError(400, `Missing role`);
                            }
                            let forecast = body["forecast"];
                            if (!forecast) {
                                forecast = 0;
                                //throw new HttpError(400, `Missing forecast this month`);
                            }
                            console.log(`➡️ POST /api/projects: assignconsultant request, projectName=${projectName}, consultantName=${consultantName}, role=${role}, forecast=${forecast}`);
                            const result = await ProjectApiService.addConsultantToProject
                                (projectName, consultantName, role, forecast);

                            res.jsonBody.results = {
                                status: 200,
                                clientName: result.clientName,
                                projectName: result.projectName,
                                consultantName: result.consultantName,
                                remainingForecast: result.remainingForecast,
                                message: result.message
                            };

                            console.log(`   ✅ POST /api/projects: response status ${res.status} - ${result.message}`);
                        } else {
                            throw new HttpError(400, `Missing request body`);
                        }
                        return res;
                    }
                    default: {
                        throw new HttpError(400, `Invalid command: ${id}`);
                    }
                }

            }
            default: {
                throw new Error(`Method not allowed: ${req.method}`);
            }
        }

    } catch (error) {

        const status = <number>error.status || <number>error.response?.status || 500;
        console.log(`   ⛔ Returning error status code ${status}: ${error.message}`);

        res.status = status;
        res.jsonBody.results = {
            status: status,
            message: error.message
        };
        return res;
    }
}

app.http("projects", {
    methods: ["GET", "POST"],
    authLevel: "anonymous",
    route: "projects/{*id}",
    handler: projects,
});

File: samples/cext-trey-research-python/functions/projects/projects.py
import azure.functions as func
import logging
import json
import re
from services.project_api_service import project_api_service
from services.identity_service import identity_service
from services.utilities import HttpError, clean_up_parameter

async def projects(req: func.HttpRequest) -> func.HttpResponse:
    logging.info("HTTP trigger function 'projects' processed a request.")

    res = {
        "status": 200,
        "results": []
    }

    try:
        # Will throw an exception if the request is not valid
        user_info = await identity_service.validate_request(req)

        # Extract the 'id' parameter from the route
        project_id = req.route_params.get('id', '').lower()

        if req.method == 'GET':
            project_name = req.params.get('projectName', '').lower()
            consultant_name = req.params.get('consultantName', '').lower()

            logging.info(f"➡️ GET /api/projects: request for projectName={project_name}, consultantName={consultant_name}, id={project_id}")

            project_name = clean_up_parameter('projectName', project_name)
            consultant_name = clean_up_parameter('consultantName', consultant_name)

            if project_id:
                result = await project_api_service.get_api_project_by_id(project_id)
                
                # Convert the ApiProject object to a dictionary
                res['results'] = [result.to_dict()] if hasattr(result, 'to_dict') else [vars(result)]
                
                logging.info(f"✅ GET /api/projects: response status {res['status']}; 1 project returned")
                return func.HttpResponse(
                    json.dumps(res),
                    status_code=res['status'],
                    mimetype="application/json"
                )

            # Fetch projects for the current user if projectName includes 'user_profile'
            if 'user_profile' in project_name:
                result = await project_api_service.get_api_projects("", user_info['name'])
                res['results'] = [result.to_dict()] if hasattr(result, 'to_dict') else [vars(result)]
                logging.info(f"✅ GET /api/projects for current user response status {res['status']}; {len(result)} projects returned")
                return func.HttpResponse(
                    json.dumps(res),
                    status_code=res['status'],
                    mimetype="application/json"
                )

            # Fetch by projectName and consultantName
            result = await project_api_service.get_api_projects(project_name, consultant_name)
            
            # Convert the list of ApiProject objects to dictionaries
            res['results'] = [r.to_dict() if hasattr(r, 'to_dict') else vars(r) for r in result]

            logging.info(f"✅ GET /api/projects: response status {res['status']}; {len(result)} projects returned")
            return func.HttpResponse(
                json.dumps(res),
                status_code=res['status'],
                mimetype="application/json"
            )

        elif req.method == 'POST':
            if project_id == 'assignconsultant':
                try:
                    raw_body = req.get_body().decode('utf-8')
                    formatted_body = re.sub(r'([a-zA-Z0-9_]+):', r'"\1":', raw_body)
                    body = json.loads(formatted_body)
                except ValueError:
                    raise HttpError(400, "No body to process this request.")

                if body:
                    project_name = clean_up_parameter('projectName', body.get('projectName'))
                    if not project_name:
                        raise HttpError(400, "Missing project name")

                    consultant_name = clean_up_parameter('consultantName', body.get('consultantName', '').lower())
                    if not consultant_name:
                        raise HttpError(400, "Missing consultant name")

                    role = clean_up_parameter('Role', body.get('role'))
                    if not role:
                        raise HttpError(400, "Missing role")

                    forecast = body.get('forecast', 0)

                    logging.info(f"➡️ POST /api/projects: assignconsultant request, projectName={project_name}, consultantName={consultant_name}, role={role}, forecast={forecast}")
                    result = await project_api_service.add_consultant_to_project(project_name, consultant_name, role, forecast)

                    res['results'] = {
                        'status': 200,
                        'clientName': result.clientName,
                        'projectName': result.projectName,
                        'consultantName': result.consultantName,
                        'remainingForecast': result.remainingForecast,
                        'message': result.message
                    }

                    logging.info(f"✅ POST /api/projects: response status {res['status']} - {result.message}")
                    return func.HttpResponse(
                        json.dumps(res),
                        status_code=res['status'],
                        mimetype="application/json"
                    )
                else:
                    raise HttpError(400, "Missing request body")
            else:
                raise HttpError(400, f"Invalid command: {project_id}")

        else:
            raise HttpError(405, f"Method not allowed: {req.method}")

    except HttpError as error:
        status = error.status or 500
        logging.error(f"⛔ Returning error status code {status}: {str(error)}")
        res['status'] = status
        res['results'] = {
            'status': status,
            'message': str(error)
        }
        return func.HttpResponse(
            json.dumps(res),
            status_code=status,
            mimetype="application/json"
        )